File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/backtrack/P46_Permutations.java
package com.dsa.leetcode.labuladong.backtrack;

import java.util.LinkedList;
import java.util.List;

/**
 * @since 2023/11/18 12:46
 */
public class P46_Permutations {
    List<List<Integer>> ans = new LinkedList<>();
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        backtrack(nums, used, 0);
        return ans;
    }

    private void backtrack(int[] nums, boolean[] used, int deep) {
        if (deep == nums.length) {
            ans.add(new LinkedList<>(path));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                used[i] = true;
                path.add(nums[i]);
                backtrack(nums, used, deep + 1);
                path.removeLast();
                used[i] = false;
            }
        }
    }

    public static void main(String[] args) {
        P46_Permutations p46Permutations = new P46_Permutations();
        System.out.println(p46Permutations.permute(new int[]{1, 2, 3}));
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/backtrack/P46_Permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans, path = [], []
        used = [False for i in range(len(nums))]
        self.backtrack(nums, 0, used, ans, path)
        return ans

    def backtrack(self, nums, deep, used, ans, path):
        if deep == len(nums):
            ans.append(path[:])
            return
        for i in range(len(nums)):
            if not used[i]:
                used[i] = True
                path.append(nums[i])
                self.backtrack(nums, deep + 1, used, ans, path)
                path.pop()
                used[i] = False


if __name__ == '__main__':
    solu = Solution()
    print(solu.permute([1, 2, 3]))
